Track registered J2I thunks on the JITServer

dmitry-ten opened the pull request:
Previously, whenever a J2I thunk was needed,
JITServer would always create a new one and send it
to the client for registration with the VM.
However, if the thunk has already been registered, we would
just register it again and waste a remote message.

This PR adds tracking of registered thunks on the server per-client.
If a thunk is already registered, do not regenerate it on the server.

dmitry-ten commented:
[Tests](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/2945/).
@mpirvu could you review?

mpirvu commented on runtime/compiler/env/VMJ9Server.cpp:
```diff
@@ -1179,17 +1179,37 @@ TR_J9ServerVM::isClassVisible(TR_OpaqueClassBlock *sourceClass, TR_OpaqueClassBl
+      auto &thunkSet = _compInfoPT->getClientData()->getRegisteredJ2IThunkSet();
+      std::string signature(signatureChars, signatureLength);
+      if (thunkSet.find(signature) != thunkSet.end())
+         return reinterpret_cast<void *>(-1);
```
Is it at all possible that this may cause a crash because we provided a fake thunk address and the compiler (at the server) could try to peek inside?

mpirvu commented on runtime/compiler/env/VMJ9Server.cpp:
```diff
@@ -1179,17 +1179,37 @@ TR_J9ServerVM::isClassVisible(TR_OpaqueClassBlock *sourceClass, TR_OpaqueClassBl
 TR_J9ServerVM::setJ2IThunk(char *signatureChars, uint32_t signatureLength, void *thunkptr, TR::Compilation *comp)
    {
-   TR_J9VMBase::setJ2IThunk(signatureChars, signatureLength, thunkptr, comp);
+   // Serialize thunk and send it to the client to be registered with the VM.
```
A potentially problematic situation due to a race:
1) Thread 1 at server does `setJ2IThunk`, registers the thunk locally in the set and sends a message to the client
2) The client keeps this thunk in its local list, to be registered at the very end of the compilation
3) Thread 2 at the server does `getJ2IThunk`, finds the thunk in the set, continues and finishes the compilation
4) The client tries to relocate the thunk for thread 2, searches the VM, but there is nothing to be found.

dmitry-ten replied:
I see the problem.
We could have a per-compilation cache of thunks that were generated but not registered by the client yet. At the end of the compilation, everything from per-compilation cache would be committed to persistent cache.
Then `getJ2IThunk` could access only the persistent cache.

dmitry-ten commented:
@mpirvu ready for another review. 
@dchopra001 FYI.

dmitry-ten commented:
Ok, nvm, I found some more bugs, fixing them now

mpirvu commented on runtime/compiler/control/JITaaSCompilationThread.cpp:
```diff
@@ -822,43 +833,80 @@ bool handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe)
+            // Do not need relocation here, because helper address should have been originally
+            // fetched from the client.
+            thunkStart = TR_JITaaSRelocationRuntime::copyDataToCodeCache(serializedThunk.data(), serializedThunk.size(), fe);
+            if (!thunkStart)
```
we could fail this compilation, but we have added an entry to our set of thunks. better to add after we receive conformation that the registration was successful at the client.

mpirvu commented on runtime/compiler/control/JITaaSCompilationThread.cpp:
```diff
@@ -822,43 +833,80 @@ bool handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe)
+
+         void *thunkAddress = reinterpret_cast<void *>(thunkStart);
+         fe->setInvokeExactJ2IThunk(thunkAddress, comp);
+         client->write(response, thunkAddress);
```
I see that the client sends back `thunkAddress` but the server reads `void`

mpirvu commented on runtime/compiler/control/JITaaSCompilationThread.cpp:
```diff
@@ -822,43 +833,80 @@ bool handleServerMessage(JITServer::ClientStream *client, TR_J9VM *fe)
-         client->write(response, JITServer::Void());
+         else
+            {
+            thunkStart = reinterpret_cast<uint8_t *>(&serializedThunk[0]);
```
I remember that invokeExact thunks are always allocated from code cache, so we must copyDataToCodeCache in this case as well. Using the address from the serialized thunk will not work.

dmitry-ten replied:
I guess we should copy it to code cache for consistency, but in AOT, this thunk will be immediately stored to SCC, so it doesn't matter where it's originally allocated.

dsouzai replied:
> but in AOT, this thunk will be immediately stored to SCC

MethodHandles aren't supported in AOT.

dmitry-ten replied:
Yeah, but why do we still have relocations and SCC stores for invoke exact thunks?

dsouzai replied:
> Yeah, but why do we still have relocations and SCC stores for invoke exact thunks?

I have no idea, but it would never work as the codebase stands; Method Handle support has yet to be added. Actually if you look for `TR_J2IThunks` in `RelocationRecord.cpp` there is no associated ReloRecord for it. Grepping for it in the codebase shows some places adding an external relocation, but those code paths are never executed in AOT lol. 

As for `TR_Thunks` and `TR_J2IVirtualThunkPointer`, they're used to relocate a Snippet which has a pointer to a J2I thunk; which is different from an Invoke* thunk (which has to do with Method Handles).

dmitry-ten replied:
So do you think we should keep doing whatever mainline does, assuming MethodHandle support will be eventually enabled for AOT, or just ignore it for now?

dsouzai replied:
> So do you think we should keep doing whatever mainline does

Yes, however, when the compiler on the server goes to register the thunks, you're going to need to:

1. send the thunk over to the client
2. have the client allocate some code cache memory and memcopy the thunk
3. register the thunk
4. send the thunk pointer (which is valid in the client's address space) over to the server

I think you have to do this for J2I and InvokeExactJ2I thunks.

dmitry-ten replied:
We already do that for non-AOT, excluding sending client's invokeExactJ2IThunk pointer back to the server, because the server will never use it anyway. 
But mainline AOT does not register the thunk, it only saves it to SCC, so wouldn't we want to keep behaviour identical?

dsouzai replied:
> But mainline AOT does not register the thunk, it only saves it to SCC, so wouldn't we want to keep behaviour identical?

Yeah it does, assuming that when server stores to the SCC it results in it getting stored to the client's SCC.

dmitry-ten replied:
Yes, it's stored to client's SCC.

dmitry-ten commented on runtime/compiler/env/VMJ9Server.cpp:
```diff
@@ -1368,7 +1401,14 @@ TR_J9ServerVM::isClassLoadedBySystemClassLoader(TR_OpaqueClassBlock *clazz)
+      // Add terse signature of the thunk to the set of registered thunks
+      OMR::CriticalSection thunkMonitor(_compInfoPT->getClientData()->getThunkSetMonitor());
+      TR_J2IThunk *thunk = reinterpret_cast<TR_J2IThunk *>(thunkptr);
+      std::string signature(thunk->terseSignature(), strlen(thunk->terseSignature()));
```
You mean you don't want to invoke `strlen`? Yes, it should work, because terse signature is guaranteed to be null-terminated.

dmitry-ten commented on runtime/compiler/runtime/RelocationRuntime.cpp:
```diff
@@ -1434,3 +1426,30 @@ TR_JITaaSRelocationRuntime::allocateSpaceInDataCache(uintptr_t metaDataSize,
+   {
+   TR::CompilationInfoPerThreadBase *compInfoPT = fe->_compInfoPT;
+   int32_t numReserved;
+   TR::CodeCache *codeCache = compInfoPT->getCompilation()->cg()->getCodeCache();
```
So the thing is, `RelocationRuntime` doesn't have a pointer to a valid code cache until `prepareRelocateAOTCodeAndData` is called. At that point, code cache is set to `cg()->getCodeCache()`, which will be also NULL on the client, so we would allocate a new code cache there.
I decided to change the code for `copyDataToCodeCache`, so that whenever we allocate code cache there, we set `cg()->setCodeCache` so that at the end of the compilation we can use the same code cache for the rest of the code received from the server.
So with my latest change, both cache pointers should point to the same cache.

mpirvu replied:
There is a danger that, when trying to allocate code for the compiled body there isn't sufficient space because, during reservation we only requested that the size of the thunks be available. It's better to let the client reserve again a new code cache that is appropriate in size.

dmitry-ten commented:
I think I addressed all of the comments, also fixed a couple of not mentioned bugs, including one not directly related to this PR, but it still affects method handles, so I decided to put it here anyway.
Started the tests, will report the results.

dmitry-ten commented:
Ok, so the tests that are supposed to pass [passed](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/2974).
@mpirvu could you give it another look?
@dchopra001 could you confirm that the issue you were facing is fixed now?

dchopra001 commented:
@dmitry-ten I'm seeing a hang on zLinux when running `make _jit_jar_0`. Without this commit, the test crashes exposing the original problem that this PR is trying to address.

dchopra001 commented:
I'm going to rescind my comment from last night. It seems like this test just takes a really long time to run. I've kicked off another run and will wait a little bit longer for it to finish. So far it looks like it will pass though.

dmitry-ten commented:
Addressed comments and modified `copyDataToCodeCache` to reserve a new code cache every time, to allocate in the cold segment, disabled contiguous allocation and method header needed, added a check for whether a compilation has been interrupted during code cache reservation.
@mpirvu Am I missing anything else?
Started [tests](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/3005/).

mpirvu commented on runtime/compiler/runtime/RelocationRuntime.cpp:
```diff
@@ -1435,3 +1427,33 @@ TR_JITaaSRelocationRuntime::allocateSpaceInDataCache(uintptr_t metaDataSize,
+   manager->allocateCodeMemory(0, totalSize, &codeCache, &coldCodeStart, false, false);
+   if (coldCodeStart)
+      {
+      memcpy(coldCodeStart, startAddress, totalSize);
```
We have to unreserve even when the allocation was successful, because the next thunk will reserve again.
Also, let's put an assert that when we enter this method, no code cache is reserved.

dmitry-ten replied:
Done.

mpirvu commented on runtime/compiler/runtime/RelocationRuntime.cpp:
```diff
@@ -1435,3 +1427,33 @@ TR_JITaaSRelocationRuntime::allocateSpaceInDataCache(uintptr_t metaDataSize,
+   if (compInfoPT->getCompThreadId() >= 0 && fe->getCompilationShouldBeInterruptedFlag())
+      {
+      codeCache->unreserve();
+      return NULL;
```
In this case we would do `failCompilation<TR::CodeCacheError>` which is not ideal because the real reason is compilation interrupted. I don't see an easy solution besides passing back an error code.

dmitry-ten replied:
I see that `TR_SharedCacheRelocationRuntime::allocateSpaceInCodeCache` used to pass back error code for that case, but it is commented out now. So mainline code has the same issue currently. It's probably fine then if we do the same.

mpirvu replied:
ok

mpirvu commented:
I have run this code and the code cache allocation issues are no longer seen. Note that my code included the suggestion to always unreserve the code cache after a thunk allocation. The number of messages related to thunks is also drastically reduced.

dmitry-ten commented:
@mpirvu added code for unreserving code cache, the tests that I started previously passed.
Should I start another build for the slightly updated code? I don't think the result will be different though.

mpirvu commented on runtime/compiler/runtime/RelocationRuntime.cpp:
```diff
@@ -1435,3 +1427,32 @@ TR_JITaaSRelocationRuntime::allocateSpaceInDataCache(uintptr_t metaDataSize,
+   int32_t numReserved;
+   TR::CodeCache *codeCache = NULL;
+   TR::CodeCacheManager *manager = TR::CodeCacheManager::instance();
+   codeCache = manager->reserveCodeCache(false, totalSize, compInfoPT->getCompThreadId(), &numReserved);
```
`numReserved` may include reservations done by other threads in system which is possible.
What we want to assert that the current thread does not have anything reserved. I believe that can be tested with `compiler->cg()->getCodeCache()`. If this is NULL, then no code cache is reserved by current compilation.

dmitry-ten replied:
ok, done.

mpirvu replied:
One more request: please put the assert before `manager->reserveCodeCache`, otherwise people may get confused as to why there should be no code cache reserved when I just reserved one.

dmitry-ten replied:
done.
